Lesson19/Merve/script.js

```diff
@@ -0,0 +1,270 @@
+        return true;
+    }
+
+    return findSpecificValue(arr.slice(1), valueToCheck);
+}
+
+console.log(findSpecificValue(numbers, valueToCheck));
+
+// 6. Write a recursive function to flatten an object with nested objects into a single-level object.
+const nestedObj = {
+	name: 'John',
+	age: 30,
+	address: {
+		street: '123 Main St',
+		city: 'New York',
+		country: 'USA',
+	},
+	hobbies: ['reading', 'coding'],
+	social: {
+		twitter: '@johndoe',
+		linkedIn: 'johndoe',
+	},
+};
+   
+   function flattenObject(obj) {
+    let flatObj = {};
+
+    for (let key in obj) {
+        if (typeof obj[key] === 'object' && !Array.isArray(obj[key])) {
+            const nestedFlatObj = flattenObject(obj[key]);
+            for (let nestedKey in nestedFlatObj) {
+                flatObj[key + '.' + nestedKey] = nestedFlatObj[nestedKey];
+            }
+        } else {
+            flatObj[key] = obj[key];
+        }
+    }
+    return flatObj;
+}
+console.log(flattenObject(nestedObj));
+
+
+// 7. Write a recursive function to find the maximum depth of a nested object.
+const nestedObjEx7 = {
+	a: 1,
+	b: {
+		c: 2,
+		d: {
+			e: 3,
+			f: {
+				g: 4,
+			},
+		},
+	},
+	h: 5,
+};
+function getMaxDepth(obj) {
+    let maxDepth = 0;
+    for (let key in obj) {
+        if (obj.hasOwnProperty(key)) {
+            const depth = getMaxDepth(obj[key]);
+            if (depth > maxDepth) {
+                maxDepth = depth;
+            }
+        }
+    }
+    return maxDepth +1;
+}
+console.log(getMaxDepth(nestedObjEx7));
+
+// 8. Write a recursive function to reverse the order of words in a sentence.
+const sentenceEx8 = 'Hello, how are you?';
+function reverseWords(sentence) {
+   function reverseArray(arr) {
+        if (arr.length <= 1) {
+            return arr;
+        }
+        return [arr.pop()].concat(reverseArray(arr));
+    }
+    const words = sentence.split(' ');
+   
+    const reversedWords = reverseArray(words);
+   
+    return reversedWords.join(' ');
+}
+
+console.log(reverseWords(sentenceEx8)); 
+
+// 9. Write a recursive function to find the length of the longest word in a sentence.
+const sentenceEx9 = 'The quick brown fox jumps over the lazy dog';
+function findLongestWordLength(sentence) {
+    const words = sentence.split(' ');
+
+    function findMaxLength(words, index) {
+        if (index === words.length) {
+            return 0;
+        }
+        const currentWordLength = words[index].length;
+        const maxLengthInRemainingWords = findMaxLength(words, index + 1);
+        return Math.max(currentWordLength, maxLengthInRemainingWords);
+    }
+    return findMaxLength(words, 0);
+}
+
+console.log(findLongestWordLength(sentenceEx9)); 
+
+// 10. Write a recursive function to check if an object contains a specified property.
+const person = {
+	name: 'John',
+	age: 30,
+	address: {
+		city: 'New York',
+		country: 'USA',
+	},
+};
+
+function findProperty(object, property){
+    for( let key in object){
```

Comment:
This solution has to be a bit different as object may contain values of type object (nested objects):
```
for (let key in object) {
        if (key === property) {
            return true;
        }
        if (typeof object[key] === 'object' && object[key] !== null) {
            if (findProperty(object[key], property)) {
                return true;
            }
        }
    }
    return false;
```